### PR TITLE
Update format lambda for new automation changes

### PR DESCRIPTION
**Pull request by hakbailey**
### What does this PR do?

Updates the format lambda to create the input JSON needed for each step of the TIMDEX pipeline. Also adds a pre-transform step for Alma that extracts the zipped export files from the Alma SFTP bucket to the TIMDEX pipeline bucket.

### Helpful background context

Although this doesn't yet handle deleted records, the ability to process multiple files during the transform and output stages sets us up to be able to add that when we have the components ready.

### How can a reviewer manually see the effects of these changes?

Run a lambda function test in Dev1 with the expected input (per the README) and confirm it generates the expected output. Note: using source Alma starting with the transform step will allow you to see it also include the Alma-specific pre-transform piece, but you'll need to use run-date=2022-10-11 as that's the only Alma export file we have in dev currently.

### Includes new or updated dependencies?

YES

### What are the relevant tickets?

- https://mitlibraries.atlassian.net/browse/TIMX-112
- https://mitlibraries.atlassian.net/browse/TIMX-113

### Developer

- [x] All new ENV is documented in README
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer

- [ ] The commit message is clear and follows our guidelines (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated or is unnecessary
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

**Review by ehanson8 — APPROVED** [review]
A few comments but looks good and works as expected in the dev1 lambda tests!

File: README.md [hunk 1]
```diff
@@ -15,90 +16,60 @@ Takes input JSON (usually from EventBridge although it can be passed to a manual
...
-  - *Note*: if provided source is "aspace", a method option is passed to the harvest command (if starting at the harvest step) to ensure that we use the "get" harvest method instead of the default "list" method used for all other sources. This is required because ArchivesSpace inexplicably provides incomplete oai-pmh responses using the "list" method.
-- `time`: required. Must be in one of the formats ["yyyy-mm-dd", "yyyy-mm-ddThh:mm:ssZ"]. The provided date *minus one day* is used in the input/output file naming scheme for all steps of the pipeline, and as the harvest end date (if starting at the harvest step)
+  - *Note*: if provided source is "aspace", a method option is passed to the harvest command (if starting at the extract step) to ensure that we use the "get" harvest method instead of the default "list" method used for all other sources. This is required because ArchivesSpace inexplicably provides incomplete oai-pmh responses using the "list" method.
 
 #### Required harvest fields
```
[the remark this answers is not in the document]
Comment (hakbailey) [reply README.md:27]: That is intentional, although I agree it's a bit confusing. Extract and harvest are not the same thing. The step in the pipeline is extract...but the type of extraction that is happening here is OAI-PMH harvest. I've deliberately left the term harvest where it is only referring to OAI-PMH harvest-related things, and used extract for anything that refers to the phase of the pipeline. The long-term reason for this is that I expect we may have other types of extraction at some point (technically we already do, it just happens outside this step function with Alma) and we'll want to continue to use this pipeline, with the same overall extract step, just a different extract type which will presumably call a different fargate task. I will update this to specify OAI-PMH harvest, but I want to keep the term harvest in both the code and documentation where it's applicable.

File: lambdas/alma_prep.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+def extract_file_from_source_bucket_to_target_bucket(
+    s3_client: boto3.client,
+    source_bucket: str,
+    source_file_key: str,
+    target_bucket: str,
+    target_file_key: str,
+) -> None:
+    """Extract a single tarred file from one s3 bucket to another s3 bucket."""
+    with smart_open.open(f"s3://{source_bucket}/{source_file_key}", "rb") as tar_file:
+        file_contents = next(extract_tarfile(tar_file))
```
[the remark this answers is not in the document]
Comment (hakbailey) [reply lambdas/alma_prep.py:23]: Hmm. That was my intention with the phrase "extract a single tarred file". Do you have a suggested change that will make it more clear?

File: lambdas/alma_prep.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        "%s Alma export files found in S3 for date %s", len(alma_export_files), run_date
+    )
+    s3_client = boto3.client("s3")
+    for export_file in alma_export_files:
+        load_type, sequence = get_load_type_and_sequence_from_alma_export_filename(
+            export_file
+        )
+        extract_output_file = helpers.generate_step_output_filename(
+            load_type,
+            helpers.generate_step_output_prefix("alma", run_date, run_type, "extract"),
```
[the remark this answers is not in the document]
Comment (hakbailey) [reply lambdas/alma_prep.py:96]: I'm not quite sure what you're referring to with this comment? The step passed is `"extract"`.

[the remark this answers is not in the document]
Comment (ehanson8) [reply lambdas/alma_prep.py:98]: Ugh, disregard, this comment clearly doesn't apply here but I cannot see what I was referring to anywhere else

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ehanson8) [reply README.md]: That makes sense, thanks for clarifying!

**Pull request merged**